### ISSUE TITLE
I/O Error when running from a different directory than where the instrument file is

### ISSUE THREAD

**Issue by shervin86** [McStasMcXtrace/McCode#1032]
Dear developers and maintainers, while running some tests with the instrument file that you can find in attachment, I got a "not such file or directory" error that seem to depend on the location from which I compile and execute the simulation. Unfortunately I have not been able to investigate further to understand what causes this error. I hope you can help me.

Here's the instrument file I am using:
[ILL_H512_D22.txt](https://github.com/McStasMcXtrace/McCode/files/4832207/ILL_H512_D22.txt)

When compiling and executing from the very same directory, everything works fine:
```
mcstas -t -o ILL_H512_D22.c ILL_H512_D22.txt
gcc -O3 -I/usr/share/mcstas/2.6/libs/mcpl -o ILL_H512_D22.out ILL_H512_D22.c -lm -L/usr/share/mcstas/2.6/libs/mcpl/ -lmcpl
./ILL_H512_D22.out  lambda=4.5 -s 554321 -n 1e6 --no-output-files
```

But now, if I try to put the .c output somewhere else, I compile from the directory where the .c is,  the compilation goes fine, and I execute the simulation from that directory, but I get the error at runtime:
```
mcstas -t -o /tmp/ILL_H512_D22.c ILL_H512_D22.txt
cd /tmp/
gcc -O3 -I/usr/share/mcstas/2.6/libs/mcpl -o ILL_H512_D22-tmp.out ILL_H512_D22.c -lm -L/usr/share/mcstas/2.6/libs/mcpl/ -lmcpl
./ILL_H512_D22-tmp.out  lambda=4.5 -s 554321 -n 1e6 --no-output-files
cd -
```
Here's the error message:
```
# McStas 2.6 - Jan. 24, 2020: [pid 924] Signal 11 detected SIGSEGV (Mem Error)
# Simulation: ILL_H512_D22 (ILL_H512_D22.txt) 
# Breakpoint: Vout (Init) 0.00 % (       0.0/ 1000000.0)
# Date:      Thu Jun 25 17:25:48 2020
# Started:   Thu Jun 25 17:25:48 2020
# Last I/O Error: No such file or directory
# McStas 2.6 - Jan. 24, 2020: Simulation stop (abort).
fish: Job 3, “./ILL_H512_D22-tmp.out  lambda=…” terminated by signal SIGQUIT (Quit request from job control with core dump (^\))
```

Now, if I do as before, but I go back to where the instrument file is, I do not get the runtime error anymore:
```
mcstas -t -o /tmp/ILL_H512_D22.c ILL_H512_D22.txt
cd /tmp/
gcc -O3 -I/usr/share/mcstas/2.6/libs/mcpl -o ILL_H512_D22-tmp.out ILL_H512_D22.c -lm -L/usr/share/mcstas/2.6/libs/mcpl/ -lmcpl
cd -
/tmp/ILL_H512_D22-tmp.out  lambda=4.5 -s 554321 -n 1e6 --no-output-files
```

I have the impression that this might be a bug, but I am not sure it is not due to some poor understanding by my side.
Thanks in advance, Shervin

**Comment by willend** [comment]
Hello @shervin86,
Yes, this can probably be characterised as a bug, thanks for reporting. 

The thing is that the "usual" way of running McStas simulations implicates a user-selected "work directory" where instrument-file, any extra bits and pieces needed e.g. self-supplied components, generated c-code and generated binary all reside. Our end users in ~99% of the cases use our scripting-layer tools mcrun and mcgui to automate the process of instrument-file compilation, and these tools implement the above practice.

As I remember, an instrument generating MCPL output will embed the instrumentfile as a binary blob, and very likely that inclusion is done without checking that the actual file exists, based upon the above silent assumption that all of the bits and pieces reside alongside each other. :-)

I don't have time until next week to dig any further, but if you feel like hunting the above bug, my feeling is that the inclusion is done directly from the MCPL_output component. (According to individual preference, this can then be called many things: missing documentation, sloppy coding, poor error-handling practice or non-standard user practice etc. :-) )

If the bug is in fact caused by the above situation, my personal preferences would be to issue a warning that the instrument file could not be found / embedded, but otherwise let the simulation continue as usual. That being said, the best would indeed be that the MCPL output can somehow be traced back to the original instrument file...

Best,
Peter

**Comment by willend** [comment]
Hi @shervin86,

Now back from vacation I tried reproducing your issue, which is funnily enough not reproduced on macOS, but moving to a Ubuntu 20.04 VM I do reproduce the issue.

It turns out the issue is glibc specific, https://stackoverflow.com/questions/16922871/why-glibcs-fclosenull-cause-segmentation-fault-instead-of-returning-error - an ```fclose(NULL)``` is done in ```MCPL_output.comp``` when the ```mcinstrument_source``` is not found.

I have improved this on the master branch in 11fb077f0 so that this should not occur in future McStas releases.
